src/Components/Web.JS/src/Platform/Circuits/CircuitManager.ts

```diff
@@ -1,22 +1,288 @@
+    throw new Error('descriptor must be defined when using a descriptor.');
+  }
+
+  if (sequence === undefined) {
+    throw new Error('sequence must be defined when using a descriptor.');
+  }
+
+  const [parsedSequenceOk, parsedSequence] = getParsedSequence(sequence);
+  if (!parsedSequenceOk) {
+    throw new Error(`Error parsing the sequence '${sequence}' for component '${json}'`);
+  }
+
+  if (!prerendered) {
+    return {
+      type,
+      selector,
+      sequence: parsedSequence,
+      descriptor,
+      start,
+    };
   } else {
-    throw new Error('Circuit failed to start');
+    const end = getComponentEndComment(prerendered, iterator);
+    if (!end) {
+      throw new Error(`Could not find an end component comment for '${start}'`);
+    }
+
+    return {
+      type,
+      selector,
+      sequence: parsedSequence,
+      descriptor,
+      start,
+      prerendered,
+      end,
+    };
   }
 }
+
+function getParsedSequence(sequence: string): [boolean, number] {
+  const result = Number.parseInt(sequence);
+  return [Number.isNaN(result), result];
+}
+
+function getComponentEndComment(prerenderedId: string, iterator: ComponentCommentIterator): ChildNode | undefined {
+  while (iterator.next() && iterator.currentElement) {
+    const node = iterator.currentElement;
+    if (node.nodeType !== Node.COMMENT_NODE) {
+      continue;
+    }
+    if (!node.textContent) {
+      continue;
+    }
+
+    const definition = new RegExp(blazorCommentRegularExpression).exec(node.textContent);
+    const json = definition && definition[1];
+    if (!json) {
+      continue;
+    }
+
+    validateEndComponentPayload(json, prerenderedId);
+
+    return node;
+  }
+
+  return undefined;
+}
+
+function validateEndComponentPayload(json: string, prerenderedId: string): void {
+  const payload = JSON.parse(json);
+  if (Object.keys(payload).length !== 1) {
+    throw new Error(`Invalid end of component comment: '${json}'`);
+  }
+  const { prerendered } = payload;
+  if (!prerendered) {
+    throw new Error(`End of component comment must have a value for the prerendered property: '${json}'`);
+  }
+  if (prerendered !== prerenderedId) {
+    throw new Error(`End of component comment prerendered property must match the start comment prerender id: '${prerenderedId}', '${prerendered}'`);
+  }
+}
+
+class ComponentCommentIterator {
+
+  private childNodes: NodeListOf<ChildNode>;
+
+  private currentIndex: number;
+
+  private length: number;
+
+  public currentElement: ChildNode | undefined;
+
+  public constructor(childNodes: NodeListOf<ChildNode>) {
+    this.childNodes = childNodes;
+    this.currentIndex = -1;
+    this.length = childNodes.length;
+  }
+
+  public next(): boolean {
+    this.currentIndex++;
+    if (this.currentIndex < this.length) {
+      this.currentElement = this.childNodes[this.currentIndex];
+      return true;
+    } else {
+      this.currentElement = undefined;
+      return false;
+    }
+  }
+}
+
+
```

Comment:
I'm not really reviewing this part in detail because I'm betting it's really similar to the code we had in preveiew7. LMK if this needs a thorough look.

Reply:
Yes. I changed the comment prefix from M.A.Components to simply Blazor and cleaned up a bit iterating through the comments (by using the componentcommentiterator) but otherwise it should be pretty similar. 

I also plan to cover this with JS tests. to validate things work.